src/classes/PSC_ManageSoftCredits_CTRL.cls

```diff
@@ -201,64 +201,91 @@ public with sharing class PSC_ManageSoftCredits_CTRL {
         for (OpportunityContactRole ocr : [SELECT Id, ContactId, Contact.Name FROM OpportunityContactRole WHERE OpportunityId = :opp.Id AND IsPrimary = true]) {
             donors.put(ocr.ContactId, ocr.Contact.Name);
         }
+
         List<SObject> toDelete = listSoftCreditsForDelete;
         List<OpportunityContactRole> toUpsertContactRoles = new List<OpportunityContactRole>();
         List<Partial_Soft_Credit__c> toUpsertPartialCredits = new List<Partial_Soft_Credit__c>();
         List<SoftCredit> upsertedSoftCredits = new List<SoftCredit>();
-        for (SoftCredit sc : softCredits) {
+        Map<Integer, OpportunityContactRole> rollbackContactRole = new Map<Integer, OpportunityContactRole>();
+        Map<Integer, Partial_Soft_Credit__c> rollbackPartialSoftCredit = new Map<Integer, Partial_Soft_Credit__c>();        
+
+        for (Integer i = 0, size = softCredits.size(); i < size; i++) {
+            SoftCredit sc = softCredits[i];
+
+            rollbackContactRole.put(i, sc.contactRole == null ? null : sc.contactRole.clone(true, true, false, false));
+            rollbackPartialSoftCredit.put(i, sc.partial == null ? null : sc.partial.clone(true, true, false, false));
```

Comment:
why when cloning the psc's and ocr's, you don't want to preserve readonly fields and autonumbers?

Reply:
I believe the readonly and auto-number fields are not relevant in this method since they are not used. However, now I see a potential of the issue due to method changes (not sure if this information is applicable in triggers though). Will make the change to preserve the readonly and auto-number fields.